Fix bug preventing accurate reporting of dropped attribute count

jack-berg opened the pull request:
Fixes https://github.com/open-telemetry/opentelemetry-java/issues/7135

jack-berg commented on sdk/common/src/main/java/io/opentelemetry/sdk/internal/AttributesMap.java:
```diff
@@ -47,10 +46,10 @@ public static AttributesMap create(long capacity, int lengthLimit) {
   /** Add the attribute key value pair, applying capacity and length limits. */
   public <T> void put(AttributeKey<T> key, T value) {
```
Because `AttributeMap extends HashMap`, a caller was accidentally bypassing this method and calling `Map#put(T, V)` directly, which was preventing the application of attribute limits.

We protect against this by delegating to `HashMap` instead of extending.

trask reviewed (approved):
(I suspect extending may have been intentional as a performance optimization to avoid the 4 bytes allocated to the delegate field)

jack-berg commented:
> I suspect extending may have been intentional as a performance optimization to avoid the 4 bytes allocated to the delegate field

The thought crossed my mind. We could keep extending, and override the implementation of all unsupported map methods with versions that throw exceptions.

jack-berg commented:
> The choices are:
> - Fix this case where we bypass the limits, but continue to extend HashMap and possibly encounter similar issues in the future. Slightly less memory allocated for basically every span.
> - Delegate to HashMap instead of extending, which makes it easy to verify that no caller is bypassing the limits. Slightly more memory allocated for basically every span.

I pushed a commit which reverts to extending HashMap. I added some javadoc calling attention to the potential pitfalls, and adjusted the `put` method signature `@Override` the super method, making it less error prone.

Given that its an internal class and changing to a delegation model would increase memory allocation, continuing to extend HashMap and fixing all internal callers seems to be the conservative approach.

trask commented on sdk/common/src/main/java/io/opentelemetry/sdk/internal/AttributesMap.java:
```diff
@@ -18,6 +18,14 @@
+ * would be more appropriate to delegate. The problem with extending is that we don't enforce that
+ * all {@link HashMap} methods for reading / writing data conform to the configured attribute
+ * limits. Therefore, it's easy to accidentally call something like {@link Map#putAll(Map)} or
+ * {@link Map#put(Object, Object)} and bypass the restrictions (see <a
```
is `Map#put(Object, Object)` still a problem now that you've changed the overload to an override?

jack-berg replied:
Ah no. Will update to use `putAll` as an example.
